File: src/java/br/com/lett/crawlernode/crawlers/corecontent/brasil/BrasilIngredientesonlineCrawler.java
package br.com.lett.crawlernode.crawlers.corecontent.brasil;

import br.com.lett.crawlernode.core.models.Card;
import br.com.lett.crawlernode.core.models.Product;
import br.com.lett.crawlernode.core.models.ProductBuilder;
import br.com.lett.crawlernode.core.session.Session;
import br.com.lett.crawlernode.core.task.impl.Crawler;
import br.com.lett.crawlernode.crawlers.extractionutils.core.TrustvoxRatingCrawler;
import br.com.lett.crawlernode.util.CrawlerUtils;
import br.com.lett.crawlernode.util.Logging;
import com.google.common.collect.Sets;
import exceptions.MalformedPricingException;
import exceptions.OfferException;
import models.Offer;
import models.Offers;
import models.RatingsReviews;
import models.pricing.*;
import org.json.JSONArray;
import org.json.JSONObject;
import org.jsoup.nodes.Document;

import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class BrasilIngredientesonlineCrawler extends Crawler {
   private final String  SELLER_FULL_NAME = "Ingredientes Online";
   private Set<String> cards = Sets.newHashSet(Card.VISA.toString(), Card.MASTERCARD.toString(), Card.AMEX.toString(), Card.DINERS.toString(), Card.HIPERCARD.toString(), Card.ELO.toString());


   public BrasilIngredientesonlineCrawler(Session session) {
      super(session);
   }

   @Override
   public List<Product> extractInformation(Document doc) throws Exception {
      super.extractInformation(doc);

      List<Product> products = new ArrayList<>();

      if (isProductPage(doc)) {
         Logging.printLogDebug(logger, session, "Product page identified: " + this.session.getOriginalURL());

         String internalPid = CrawlerUtils.scrapStringSimpleInfo(doc, ".product.attribute.sku > div", true);
         String internalId = internalPid;
         String name = CrawlerUtils.scrapStringSimpleInfo(doc, ".page-title", true);
         List<String> images = scrapImages(doc);
         String primaryImage = images.isEmpty() ? "" : images.remove(0);
         String description = CrawlerUtils.scrapSimpleDescription(doc, Collections.singletonList(".product-info-description"));
         RatingsReviews ratingsReviews = scrapRating(doc);
         boolean availableToBuy = doc.selectFirst(".box-tocart") != null;
         Offers offers = availableToBuy ? scrapOffers(doc) : new Offers();

         Product product = ProductBuilder.create()
            .setUrl(session.getOriginalURL())
            .setInternalId(internalId)
            .setInternalPid(internalPid)
            .setName(name)
            .setPrimaryImage(primaryImage)
            .setSecondaryImages(images)
            .setDescription(description)
            .setOffers(offers)
            .setRatingReviews(ratingsReviews)
            .build();
         products.add(product);
      } else {
         Logging.printLogDebug(logger, session, "Not a product page " + this.session.getOriginalURL());
      }

      return products;
   }

   private Offers scrapOffers(Document doc) throws MalformedPricingException, OfferException {
      Offers offers = new Offers();
      Pricing pricing = scrapPricing(doc);
      List<String> sales = Collections.singletonList(CrawlerUtils.calculateSales(pricing));

      offers.add(Offer.OfferBuilder.create()
         .setUseSlugNameAsInternalSellerId(true)
         .setSellerFullName(SELLER_FULL_NAME)
         .setMainPagePosition(1)
         .setIsBuybox(false)
         .setIsMainRetailer(true)
         .setPricing(pricing)
         .setSales(sales)
         .build());

      return offers;
   }

   private Pricing scrapPricing(Document doc) throws MalformedPricingException {
      Double spotlightPrice = CrawlerUtils.scrapDoublePriceFromHtml(doc, ".product-info-main span[data-price-type=finalPrice]", null, false, ',', session);
      Double priceFrom = CrawlerUtils.scrapDoublePriceFromHtml(doc, ".product-info-main span[data-price-type=oldPrice]", null, false, ',', session);

      if(Objects.equals(priceFrom, spotlightPrice)) {
         priceFrom = null;
      }

      return Pricing.PricingBuilder.create()
         .setPriceFrom(priceFrom)
         .setSpotlightPrice(spotlightPrice)
         .setCreditCards(scrapCreditCards(doc, spotlightPrice))
         .setBankSlip(scrapBankSlip(doc, spotlightPrice))
         .build();
   }

   private CreditCards scrapCreditCards(Document doc, Double spotlightPrice) throws MalformedPricingException {
      CreditCards creditCards = new CreditCards();
      Installments installments = scrapInstallments(doc, spotlightPrice);
      for (String card : cards) {
         creditCards.add(
            CreditCard.CreditCardBuilder.create()
               .setBrand(card)
               .setInstallments(installments)
               .setIsShopCard(false).build());
      }
      return creditCards;
   }

   private Installments scrapInstallments(Document doc, Double spotlightPrice) throws MalformedPricingException {
      Installments installments = new Installments();
      installments.add(
         Installment.InstallmentBuilder.create()
            .setInstallmentNumber(1)
            .setInstallmentPrice(spotlightPrice).build());

      if (doc.selectFirst(".installments") != null) {
         int parcels = CrawlerUtils.scrapIntegerFromHtml(doc, ".installment_period", true, 0);
         Double price = CrawlerUtils.scrapDoublePriceFromHtml(doc, ".installment_value", null, true, ',', session);
         installments.add(Installment.InstallmentBuilder.create()
            .setInstallmentNumber(parcels)
            .setInstallmentPrice(price)
            .build());
      }
      return installments;
   }

   private BankSlip scrapBankSlip(Document doc, Double spotlightPrice) throws MalformedPricingException {
      if (doc.selectFirst(".bankslip_excerpt") != null) {
         Double price = CrawlerUtils.scrapDoublePriceFromHtml(doc, ".bankslip_excerpt .price", null, true, ',', session);
         return BankSlip.BankSlipBuilder.create().setFinalPrice(price).build();
      }
      return BankSlip.BankSlipBuilder.create().setFinalPrice(spotlightPrice).build();
   }

   private List<String> scrapImages(Document doc) {
      String imagesScript = CrawlerUtils.scrapScriptFromHtml(doc, ".product.media script");
      JSONArray imagesArray = CrawlerUtils.stringToJsonArray(imagesScript);
      JSONObject imagesJson = imagesArray.isEmpty() ? null : imagesArray.getJSONObject(0);
      if(imagesJson != null) {
         JSONArray imagesData = (JSONArray) imagesJson.optQuery("/[data-gallery-role=gallery-placeholder]/mage~1gallery~1gallery/data");
         if(imagesData != null && !imagesData.isEmpty()) {
            return IntStream.range(0, imagesData.length()).mapToObj(i -> imagesData.getJSONObject(i).optString("full"))
               .collect(Collectors.toList());
         }
      }
      return Collections.emptyList();
   }

   private RatingsReviews scrapRating(Document doc) {
      String productTrustVoxId = CrawlerUtils.scrapStringSimpleInfoByAttribute(doc, ".price-box.price-final_price", "data-product-id");
      TrustvoxRatingCrawler trustVox = new TrustvoxRatingCrawler(session, "111754", logger);
      return trustVox.extractRatingAndReviews(productTrustVoxId, doc, dataFetcher);
   }

   private boolean isProductPage(Document doc) {
      return doc.selectFirst(".product-info-main") != null;
   }
}

File: src/java/br/com/lett/crawlernode/crawlers/corecontent/brasil/BrasilIngredientesonlineCrawler.kt
package br.com.lett.crawlernode.crawlers.corecontent.brasil

import br.com.lett.crawlernode.core.fetcher.FetchMode
import br.com.lett.crawlernode.core.fetcher.models.Request
import br.com.lett.crawlernode.core.models.Card
import br.com.lett.crawlernode.core.models.Product
import br.com.lett.crawlernode.core.models.ProductBuilder
import br.com.lett.crawlernode.core.session.Session
import br.com.lett.crawlernode.core.task.impl.Crawler
import br.com.lett.crawlernode.util.*
import exceptions.MalformedPricingException
import models.Offer
import models.Offers
import models.pricing.BankSlip
import models.pricing.CreditCard.CreditCardBuilder
import models.pricing.CreditCards
import models.pricing.Installment.InstallmentBuilder
import models.pricing.Installments
import models.pricing.Pricing
import org.jsoup.Jsoup
import org.jsoup.nodes.Document
import java.util.*

class BrasilIngredientesonlineCrawler(session: Session) : Crawler(session) {
   companion object {
      const val SELLER_NAME: String = "Ingredientes Online"
   }

   init {
      config.fetcher = FetchMode.APACHE
   }

   override fun fetch(): Any? {
      val cookiesHome = dataFetcher.get(session, Request.RequestBuilder.create().setUrl("https://www.ingredientesonline.com.br/").build()).cookies
      cookies.addAll(cookiesHome)
      val request = Request.RequestBuilder.create().setCookies(cookies).setUrl(session.originalURL).build()
      val response = dataFetcher[session, request]
      val html = response.body
      return Jsoup.parse(html)
   }

   override fun extractInformation(doc: Document): MutableList<Product> {
      super.extractInformation(doc)

      if (!isProductPage(doc)) {
         return mutableListOf()
      }

      val name = CrawlerUtils.scrapStringSimpleInfo(doc, ".page-title", false)
      val internalId = CrawlerUtils.scrapStringSimpleInfo(doc, "[itemprop=sku]", false)
      val primaryImage = CrawlerUtils.scrapSimplePrimaryImage(doc, ".gallery-placeholder__image", Arrays.asList("src"), "https", "")
      val secondaryImages = doc.select(".product-view .product-image div:not(:first-child) img").map { it?.attr("src") }
      .toMutableList().filterNotNull()
      val description = CrawlerUtils.scrapStringSimpleInfo(doc, ".content .value p", false)
      val offers = scrapOffers(doc)

      val product = ProductBuilder()
      .setUrl(session.originalURL)
      .setInternalId(internalId)
      .setName(name)
      .setPrimaryImage(primaryImage)
      .setSecondaryImages(secondaryImages)
      .setDescription(description)
      .setOffers(offers)
      .build()

      return mutableListOf(product)
   }

   private fun scrapOffers(doc: Document): Offers {
      val offers = Offers()
      val re = Regex("[^0-9]")

      val buyButton = CrawlerUtils.scrapStringSimpleInfo(doc, "#product-addtocart-button span", false)?: null

      if (buyButton != null) {
         val installments = scrapInstallments(doc)

         val priceText = CrawlerUtils.scrapStringSimpleInfo(doc, ".old-price .price", false)

         var priceFrom = if (priceText == null) {
            null
         } else {
            MathUtils.parseDoubleWithComma(priceText)
         }

         var spotlightText = CrawlerUtils.scrapStringSimpleInfo(doc, ".normal-price .price", false)

         if (spotlightText == null) {
            spotlightText = CrawlerUtils.scrapStringSimpleInfo(doc, ".special-price .price", false)
         }

         var spotlightPrice = if (spotlightText.toDoubleComma()?.round() == null) {
            CrawlerUtils.scrapStringSimpleInfo(doc, ".product-info-price .price", false).toDoubleComma()?.round()
         } else {
            spotlightText.toDoubleComma()?.round()
         }

         spotlightPrice.let {
            val creditCards = CreditCards(
               listOf(Card.MASTERCARD, Card.VISA, Card.AMEX, Card.DINERS, Card.ELO,
               Card.HIPERCARD).map { card: Card ->
               try {
                  return@map CreditCardBuilder.create()
                  .setBrand(card.toString())
                  .setIsShopCard(false)
                  .setInstallments(installments)
                  .build()
               } catch (e: MalformedPricingException) {
                  throw RuntimeException(e)
               }
            })
            val bankSlipPrice = CrawlerUtils.scrapStringSimpleInfo(doc, ".bankslip_excerpt .price", false).toDoubleComma()?.round()
            val bankSlipDiscountPercentage = CrawlerUtils.scrapStringSimpleInfo(doc, ".bankslip_excerpt small", false)

            var bankSlipDiscountValue = re.replace(bankSlipDiscountPercentage.toString(), "").toDouble()

            var bankSlip: BankSlip? = BankSlip.BankSlipBuilder()
            .setFinalPrice(bankSlipPrice)
            .setOnPageDiscount(((bankSlipDiscountValue) / 100)) // convert from int to decimal percent
            .build()

            if ((bankSlipPrice ?: 0.0) <= 0) {
               bankSlip = null
            }
               offers.add(
               Offer.OfferBuilder.create()
               .setPricing(
                  Pricing.PricingBuilder.create()
                  .setCreditCards(creditCards)
                  .setSpotlightPrice(spotlightPrice)
                  .setBankSlip(bankSlip)
                  .setPriceFrom(priceFrom)
                  .build()
               )
               .setSales(listOf())
               .setIsMainRetailer(true)
               .setIsBuybox(false)
               .setUseSlugNameAsInternalSellerId(true)
               .setSellerFullName(SELLER_NAME)
               .build()
            )
         }
      }
      return offers
   }

   private fun scrapInstallments(doc: Document): Installments {
      val installments = Installments()

      val parcel = CrawlerUtils.scrapStringSimpleInfo(doc, ".installment_period", false) ?: 0
      val price = CrawlerUtils.scrapStringSimpleInfo(doc, ".installment_value", false).toDoubleComma() ?: 0.0

      val re = Regex("[^0-9]")
      var parcelNumber = re.replace(parcel.toString(), "").toInt()

      val installment = InstallmentBuilder()
      .setInstallmentNumber(parcelNumber)
      .setInstallmentPrice(price.round())
      .setFinalPrice(parcelNumber * price)
      .build()

      installments.add(installment)

      return installments
   }

   private fun isProductPage(doc: Document): Boolean {
      return CrawlerUtils.scrapStringSimpleInfo(doc, ".product-name", false) != null
   }
}
